sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,240 @@
+template <typename T>
+class
+#if defined(__SYCL_DEVICE_ONLY__) || defined(__INTEL_SYCL_HOST_COMPILER__)
+  [[__sycl_detail__::uniquely_identifiable_object("specialization_id")]]
+#endif
+specialization_id {
+// ...
+};
+```
+
+## Common runtime part
+
+The runtime should implement the following function, which will be called from
+a code generated by the compiler (see the next section):
+
+```
+void __register_uniquely_identifiable_object(
+  void *Address, const char* UniqueID, const char *Kind);
+```
+
+The function accepts the following arguments:
+- `Address` is an address of a variable, which exists in an application on host
+- `UniqueID` is a unique symbolic ID, which corresponds to that variable
+- `Kind` is a string which corresponds to `kind` argument passed to
+  `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute attached
+  to the type of the variable identified by `Address`. It can be used to
+  distinguish different entities like `specialization_id` and `device_global`:
+  for example they could be stored in different maps to speed up certain
+  operations with them.
+
+The compiler guarantees that the function will be called zero or more times
+(depending on the amount of uniquely identifiable objects found in a program)
+_before_ application's `main()` function, i.e. in a global constructor.
+
+That poses some restrictions on those uniquely identifiable object, i.e. that
+they can't be used from another global object due to risk of accessing a
+non-initialized object, but that is an UB anyway because the order of global
+objects initialization is not defined in C++ when those objects are defined in
+different translation unit.
+
+## Compiler driver part
+
+The compiler driver is the component which is responsible for selecting the
+approach we are taking and the decision is made based on whether or not
+3rd-party host compiler is in use.
+
+If `-fsycl-host-compiler` option is present, the compiler driver chooses the
+integration footer approach:
+- it supplies device compilation step with `-fsycl-int-footer` option to
+  instruct device compiler to emit integration footer
+- it appends the integration footer to user-provided translation unit before
+  passing it to a host compiler
+
+Otherwise, if `-fsycl-host-compiler` is not present, then the compiler driver
+chooses another approach by simply doing nothing related to integration footer:
+- `-fsycl-int-footer` is **not** passed to device compiler
+- user-provided translation unit is passes as-is to host compiler
+
+## Integration footer approach
+
+When this approach is used, not only extra file (integration footer) is
+generated, but integration header is also modified: FE compiler generates a
+definition of a namespace scope variable of type
+`__sycl_device_global_registration` whose sole purpose it to run its constructor
+before the application's `main()` function:
+
+```
+namespace sycl::detail {
+namespace {
+
+class __sycl_device_global_registration {
```

Comment:
Did I get it right that presence of a variable whose type is marked with `[[__sycl_detail__::uniquely_identifiable_object(kind)]` attribute triggers emission of this content in integration footer and header?
Right now presence of `device_global` and specialization_id variable trigger that, `device_global` is recognized via attribute, `specialization_id` is recognized by class name.

Reply:
> Did I get it right that presence of a variable whose type is marked with [[__sycl_detail__::uniquely_identifiable_object(kind)] attribute triggers emission of this content in integration footer and header?

Right. We can even choose to always emit the registration object, but leave its constructor empty if we haven't found any uniquely identifiable objects

Reply:
This will however only happen when 3rd part host compiler is used right? For DPCPP compiler, we will need to modify current implementation to use other approach right?

Reply:
> This will however only happen when 3rd part host compiler is used right? For DPCPP compiler, we will need to modify current implementation to use other approach right?

Yes, I think that my previous comment is not entirely correct, because when DPCPP compiler is used as host as well, we are not supposed to emit integration footer at all, which means that the registration object should not be emitted as well to avoid errors about undefined functions at link time.

To summarize:
When integration footer is enabled, we emit the registration object in the header and we emit its constructor in the footer (possibly empty). When integration footer is disabled, we don't emit the registration object at all.